third_party/move/move-model/src/ty.rs

```diff
@@ -3264,6 +3264,8 @@
     pub used_modules: BTreeSet<ModuleId>,
     /// Whether to use `m::T` for representing types, for stable output in docgen
     pub use_module_qualification: bool,
+    /// Var type that should appear as Self in display
+    pub opt_self_var: Option<u32>,
```

Comment:
I don't understand this. `Self` does not play a special rule, and a variable specifically shouldn't be mapped to anything as long as it is not assigned. This does not seem to be the right fix. What is the problem? Is it a cyclic substitution? Then this is a bug around a cycle check missing in type unification.

Reply:
I've added a comment to the issue to carefully explain the problem, which is that
- To display a variable type constrained by just a `SomeReceiverFunction` constraint, we display the signature of the function, whose first parameter type is the variable type itself, so we try to display it...

`Self` does play a special role, in that we may have a type that is unconstrained other than by being the target of a receiver function call.  This happens only in the case of an error, but that's why we are displaying the type.

Possible solutions include:
1. Catching the error earlier and avoiding any attempts to print out the `SomeReceverFunction` constraint.
2. Always displaying the first parameter type of `SomeReceiverFunction` as `Self`, ignoring the actual type.
    - This seems like it could lose info in some other cases, where the first type is displayable without this loop
3. Adding a hack to show `Self` in this particular case.
4. Generalize this hack to be able to print recursive types if they happen to show up:  Add new substitution for the variable, and if the substitution is used, then display an explicitly recursive type such as `exists v. fun self.method(v, ...)`

I was able to get rid of one occurrence of the error earlier, but one remained, so I went to solution #3.  I alluded to the possible desire to generalize to solution #4 in the PR summary, but I think it's overkill for the current situation.

Reply:
I've changed it to use `..` instead of `Self`, and also to allow multiple variables to be stored when displaying via constraints to catch potential recursions.  I also check the error type name for `*error*` in the one case where it is showing up, and avoid showing the error at all in this case, since it is guaranteed to be redundant.  This has the unfortunate side-effect of leaving us without any test case to show recursion happening, but it was tested before this change.